packages/core/crates/core-protocol/src/ack/processor.rs

```diff
@@ -0,0 +1,327 @@
+                // No pending ticket, nothing to do.
+                debug!("received acknowledgement as sender: first relayer has processed the packet.");
+
+                #[cfg(all(feature = "prometheus", not(test)))]
+                METRIC_RECEIVED_SUCCESSFUL_ACKS.increment();
+
+                Ok(Reply::Sender(ack.ack_challenge()))
+            }
+
+            PendingAcknowledgement::WaitingAsRelayer(unacknowledged) => {
+                // Try to unlock our incentive
+                unacknowledged.verify_challenge(&ack.ack_key_share).map_err(|e| {
+                    #[cfg(all(feature = "prometheus", not(test)))]
+                    METRIC_RECEIVED_FAILED_ACKS.increment();
+
+                    AcknowledgementValidation(format!(
+                        "the acknowledgement is not sufficient to solve the embedded challenge, {e}"
+                    ))
+                })?;
+
+                self.db
+                    .read()
+                    .await
+                    .get_channel_from(&unacknowledged.signer)
+                    .await
+                    .map_err(|e| {
+                        #[cfg(all(feature = "prometheus", not(test)))]
+                        METRIC_RECEIVED_FAILED_ACKS.increment();
+
+                        AcknowledgementValidation(format!(
+                            "acknowledgement received for channel that does not exist, {e}"
+                        ))
+                    })?;
+
+                let domain_separator = self
+                    .db
+                    .read()
+                    .await
+                    .get_channels_domain_separator()
+                    .await
+                    .unwrap()
+                    .ok_or(MissingDomainSeparator)?;
+
+                let ack_ticket = unacknowledged.acknowledge(&ack.ack_key_share, &self.chain_key, &domain_separator)?;
+
+                // replace the un-acked ticket with acked ticket.
+                self.db
+                    .write()
+                    .await
+                    .replace_unack_with_ack(&ack.ack_challenge(), ack_ticket.clone())
+                    .await?;
+
+                #[cfg(all(feature = "prometheus", not(test)))]
+                METRIC_ACKED_TICKETS.increment();
+
+                // Check if ticket is a win
+                if ack_ticket.is_winning_ticket(&domain_separator) {
+                    debug!("{ack_ticket} is a win");
+
+                    #[cfg(all(feature = "prometheus", not(test)))]
+                    METRIC_WINNING_TICKETS_COUNT.increment();
+                } else {
+                    warn!("encountered losing {ack_ticket}");
+                    self.db.write().await.mark_losing_acked_ticket(&ack_ticket).await?;
+
+                    #[cfg(all(feature = "prometheus", not(test)))]
+                    METRIC_LOSING_TICKETS_COUNT.increment();
+                }
+
+                Ok(Reply::RelayerWinning(ack_ticket))
```

Comment:
This does not seem right. Shouldn't it return RelayerLosing/RelayerWinning based on the condition above?

Reply:
Forgot to push the last commit.